### PR TITLE
Tutorial on updating R packages and R in a project renv environment

### PR DESCRIPTION
**Pull request by florisvdh**
## Description

See title. I applied the second part of the tutorial (R upgrade to 4.1) in a `renv` project yesterday, so I thought it would be useful to share.

## Task list

- [X] My tutorial or article is placed in a subfolder of `tutorials/content`
- [X] The filename of my tutorial or article is `index.md`. In case of an Rmarkdown tutorial I have knitted my `index.Rmd` to `index.md` (both files are pushed to the repo). 
- [X] I have included `tags` in the YAML header (see the tags listed in the [tutorials website side bar](https://inbo.github.io/tutorials/) for tags that have been used before)
- [X] I have added `categories` to the YAML header and my category tags are from the [list of category tags](https://github.com/inbo/tutorials/blob/master/static/list_of_categories)

### CODE REVIEW

File: content/tutorials/r_renv_update/index.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+```r
+renv::upgrade() # upgrades renv. See note in previous section.
+renv::hydrate(update = "all") # populates the renv cache with copies of up to 
+                              # date package versions, needed by the project
+  # renv::update() should have no effect now, but running it as well won't harm 
+  # to check that all packages are indeed up to date
+renv::snapshot() # inspect the message before confirming to overwrite renv.lock
+```
+
+If you use version control, then don't forget to effectively commit the changed `renv.lock` file.
```
Comment (hansvancalster) [inline content/tutorials/r_renv_update/index.md:73]: Probably best to write this sentence also at the end of the previous section.

**Review by hansvancalster — APPROVED** [review]
Looks good! Great tutorial 👍

File: content/tutorials/r_renv_update/index.md (hunk 1 above)
Comment (florisvdh) [reply content/tutorials/r_renv_update/index.md:75]: Addressed by 862c0ee4

File: content/tutorials/r_renv_update/index.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+categories: ["r"]
+tags: ["renv", "r", "packages"]
+---
+
+## What is `renv`?
+
+`renv` is an R package that provides tools to save the version of R and all packages used in a particular directory (e.g. an RStudio project directory), and to reproduce (recreate) this environment on another (or your future) system.
+Because software versions can have important effects on the results of your work, it is important to be able to reproduce versions.
+
+For basic advice on how to setup and use `renv`, see <https://rstudio.github.io/renv>.
```
Comment (ThierryO) [inline content/tutorials/r_renv_update/index.md:15]: Maybe highlight this section by adding a header. So people can easily find instructions on how to start with `renv`

File: content/tutorials/r_renv_update/index.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+1. Update all installed packages in the default user R library of your system, i.e. outside an `renv` environment.
+See [this](../../installation/user/user_install_r) tutorial.
+In this way these package versions will need to be downloaded and installed only once.
+You may also want to update packages from other sources, such as GitHub: use `remotes::install_github()` to install the version you want.
+1. Update packages in a project `renv` environment, and store the new state in `renv.lock`.
+
+## Procedure
+
+**We assume you already applied step 1 above!**
```
Comment (ThierryO) [inline content/tutorials/r_renv_update/index.md:33]: I find it a bit confusing to have an important step of the procedure explained in detail outstide of the procedure. Maybe add the full detail here and only add a summary under Aim.

File: content/tutorials/r_renv_update/index.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+```r
+renv::upgrade() # upgrades renv, if new version is available
+renv::update() # updates packages from CRAN and GitHub, within the project
+renv::snapshot() # inspect the message before confirming to overwrite renv.lock
+```
+
+Notes:
+
+- `renv` typically already performs an automatic upgrade (`renv::upgrade()`) upon launching the `renv` environment (reported on launch).
+- To update packages in the `renv` environment, `renv` will take package copies of your user R library if the right versions are available there.
```
Comment (ThierryO) [inline content/tutorials/r_renv_update/index.md:48]:
```suggestion
- To update packages in the `renv` environment, `renv` will take package copies of your user R library if the required versions are available there.
```

File: content/tutorials/r_renv_update/index.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+Note that the copies are stored in a [global `renv` package cache](https://rstudio.github.io/renv/articles/renv.html#cache-1), i.e. a central directory where `renv` stores packages versions for all projects where it is used, for each R `x.y` version, on an as-needed basis.
+The project environment itself is based on symlinks to the particular package versions in the cache.
+
+If you use version control, then don't forget to effectively commit the changed `renv.lock` file.
```
Comment (ThierryO) [inline content/tutorials/r_renv_update/index.md:52]: Should we list which `renv` files to under version control and which you shouldn't? We could add does to the [.gitignore](https://github.com/inbo/checklist/blob/main/inst/generic_template/gitignore) template in checklist

File: content/tutorials/r_renv_update/index.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+### Updating R and packages
+
+This procedure is relevant to the `x.y` upgrades, not to the patches.
+That is, when `x` or `y` change in R version `x.y.z`.
+When only `z` has changed, previous procedure still applies.
```
Comment (ThierryO) [inline content/tutorials/r_renv_update/index.md:58]:
```suggestion
When only `z` has changed (i.e. a 'patch'), previous procedure still applies. 
```

**Review by ThierryO — APPROVED** [review]
Nice tutorial. Just a few suggestions.

File: content/tutorials/r_renv_update/index.md (hunk 2 above)
Comment (florisvdh) [reply content/tutorials/r_renv_update/index.md:15]: :+1:

File: content/tutorials/r_renv_update/index.md (hunk 3 above)
Comment (florisvdh) [reply content/tutorials/r_renv_update/index.md:33]: OK

File: content/tutorials/r_renv_update/index.md (hunk 5 above)
Comment (florisvdh) [reply content/tutorials/r_renv_update/index.md:52]: This is automatically taken care of by `renv::init()`, it creates `renv/.gitignore`. While `renv.lock` is essential, all new (untracked) files are actually relevant for git I believe, for an optimal reproducible (interactive) environment. As an example from inbo/n2khab-preprocessing:

```bash
$ git log --name-status 30faabf^..30faabf
commit 30faabfd9518dbed8402d95f791ee25ddad760fb
Author: florisvdh <floris.vanderhaeghe@inbo.be>
Date:   Mon Nov 30 15:05:15 2020 +0100

    Generate watercourse_100m...: use renv

A       src/generate_watercourse_100mseg/.Rprofile
A       src/generate_watercourse_100mseg/renv.lock
A       src/generate_watercourse_100mseg/renv/.gitignore
A       src/generate_watercourse_100mseg/renv/activate.R
A       src/generate_watercourse_100mseg/renv/settings.dcf
```

I'll rework this a bit in the introduction, but will not go into detail as the focus of the tutorial is the updating aspect, not `renv` in general. When updating, `renv.lock` generally is the only thing that is updated (unless `renv::upgrade()` resulted in a new `renv/activate.R`).

Comment (florisvdh) [reply content/tutorials/r_renv_update/index.md:52]: See also item 3 under https://rstudio.github.io/renv/articles/collaborating.html#collaborating-with-renv-1

File: content/tutorials/r_renv_update/index.md (hunk 2 above)
Comment (florisvdh) [reply content/tutorials/r_renv_update/index.md]: 0e276773

File: content/tutorials/r_renv_update/index.md (hunk 5 above)
Comment (florisvdh) [reply content/tutorials/r_renv_update/index.md:57]: 0e276773

File: content/tutorials/r_renv_update/index.md (hunk 3 above)
Comment (florisvdh) [reply content/tutorials/r_renv_update/index.md]: bdd06715

**Comment by florisvdh** [comment]
Thanks for reviewing @hansvancalster and @ThierryO.

**Pull request merged**